File: twenty_forty_eight/twenty_forty_eight.py
"""
Christian Felt
December 27, 2019
Solves the 2048 problem on Kattis.
"""


class Tile():
    """Represent a playing tile."""

    def __init__(self, number, merged=False):
        self.number = number
        self.merged = merged


def display_board(board):
    """Print the numerical values and arrangement of tiles on the board."""
    for i in range(len(board)):
        for j in range(len(board[0])):
            print(board[i][j].number, end=" ")
        print("")


def reset_merged_markers(board):
    """Set all merged instance variables to False."""
    for i in range(len(board)):
        for j in range(len(board[0])):
            board[i][j].merged = False


def do_move(move, board):
    """Do a single move of the 2048 game."""
    reset_merged_markers(board)
    if move == 0:  # Left
        did_swap = True
        while did_swap:
            did_swap = False
            for i in range(len(board)):
                for j in range(len(board[0])):
                    if j == 0:
                        pass
                    elif board[i][j - 1].number == 0 and board[i][j].number != 0:
                        board[i][j - 1] = board[i][j]
                        board[i][j] = Tile(0, False)
                        did_swap = True
                    elif board[i][j - 1].number == board[i][j].number and board[i][j - 1].merged is False and board[i][
                        j].merged is False:
                        board[i][j - 1] = Tile(board[i][j - 1].number + board[i][j].number, True)
                        board[i][j] = Tile(0, False)
                        did_swap = True
    elif move == 1:  # Up
        did_swap = True
        while did_swap:
            did_swap = False
            for j in range(len(board[0])):
                for i in range(len(board)):
                    if i == 0:
                        pass
                    elif board[i - 1][j].number == 0 and board[i][j].number != 0:
                        board[i - 1][j] = board[i][j]
                        board[i][j] = Tile(0, False)
                        did_swap = True
                    elif board[i - 1][j].number == board[i][j].number and board[i - 1][j].merged is False and board[i][
                        j].merged is False:
                        board[i - 1][j] = Tile(board[i - 1][j].number + board[i][j].number, True)
                        board[i][j] = Tile(0, False)
                        did_swap = True
    elif move == 2:  # Right
        did_swap = True
        while did_swap:
            did_swap = False
            for i in range(len(board)):
                for j in range(len(board[0]) - 1, -1, -1):
                    if j == len(board[0]) - 1:
                        pass
                    elif board[i][j + 1].number == 0 and board[i][j].number != 0:
                        board[i][j + 1] = board[i][j]
                        board[i][j] = Tile(0, False)
                        did_swap = True
                    elif board[i][j + 1].number == board[i][j].number and board[i][j + 1].merged is False and board[i][
                        j].merged is False:
                        board[i][j + 1] = Tile(board[i][j + 1].number + board[i][j].number, True)
                        board[i][j] = Tile(0, False)
                        did_swap = True
    elif move == 3:  # Down
        did_swap = True
        while did_swap:
            did_swap = False
            for j in range(len(board[0])):
                for i in range(len(board) - 1, -1, -1):
                    if i == len(board) - 1:
                        pass
                    elif board[i + 1][j].number == 0 and board[i][j].number != 0:
                        board[i + 1][j] = board[i][j]
                        board[i][j] = Tile(0, False)
                        did_swap = True
                    elif board[i + 1][j].number == board[i][j].number and board[i + 1][j].merged is False and board[i][
                        j].merged is False:
                        board[i + 1][j] = Tile(board[i + 1][j].number + board[i][j].number, True)
                        board[i][j] = Tile(0, False)
                        did_swap = True
    else:
        print("Invalid Input.")


def make_tiles(line):
    """Given a list of numbers, return a list of Tiles."""
    tiles = []
    for number in line:
        tiles.append(Tile(number, False))
    return tiles


if __name__ == "__main__":
    """Solve the 2048 problem."""
    board = []
    for i in range(4):
        board.append(make_tiles(list(map(int, input().split()))))
    move = int(input())
    do_move(move, board)
    display_board(board)

File: twenty_forty_eight/twenty_forty_eight.cpp
/*
Christian Felt
December 2019
Solves the 2048 problem on Kattis.
*/

#include <iostream>
#include <vector>
using namespace std;


class Tile
{
	/* Represent a tile. */
	public:
		bool merged; // Whether or not a tile has been merged with another tile this turn.
		int number; // The number value on the tile.
		
		Tile()
		{
			number = 0;
			merged = false;
		}
		
		Tile(int number, bool merged)
		{
			number = number;
			merged = merged;
		}
};

void display_board(vector<vector<Tile> >& board)
/* Print numbers on tiles and their arrangement on the board. */
{
	for (int i = 0; i < 4; i++)
	{
		for(int j = 0; j < 4; j++)
		{
			cout << board[i][j].number << " ";
		}
		cout << "\n";
	}
}

void reset_merged_markers(vector<vector<Tile> >& board)
/* Set all merged instance variables to false. */
{
	for (int i = 0; i < 4; i++)
	{
		for(int j = 0; j < 4; j++)
		{
			board[i][j].merged = false;
		}
	}
}

void do_move(int move, vector<vector<Tile> >& board)
/* Do a single move of the game. */
{
	reset_merged_markers(board);
	if (move == 0) // Left.
	{
		bool did_swap = true;
		while(did_swap)
		{
			did_swap = false;
			for (int i = 0; i < 4; i++)
			{
				for(int j = 0; j < 4; j++)
				{
					if (j == 0)
					{
						
					}
					else if (board[i][j-1].number == 0 && board[i][j].number != 0)
					{
						board[i][j-1] = board[i][j];
						board[i][j].number = 0;
						board[i][j].merged = false;
						did_swap = true;
					}
					else if ((board[i][j-1].number == board[i][j].number) && (board[i][j-1].merged == false && board[i][j].merged == false))
					{
						board[i][j-1].number = 2 * board[i][j-1].number;
						board[i][j-1].merged = true;
						board[i][j].number = 0;
						board[i][j].merged = false;
						did_swap = true;
					}
				}
			}
		}
	}
	else if (move == 1)  // Up
	{
		bool did_swap = true;
		while(did_swap)
		{
			did_swap = false;
			for (int j = 0; j < 4; j++)
			{
				for(int i = 0; i < 4; i++)
				{
					if (i == 0)
					{
						
					}
					else if (board[i-1][j].number == 0 && board[i][j].number != 0)
					{
						board[i-1][j] = board[i][j];
						board[i][j].number = 0;
						board[i][j].merged = false;
						did_swap = true;
					}
					else if ((board[i-1][j].number == board[i][j].number) && (board[i-1][j].merged == false && board[i][j].merged == false))
					{
						board[i-1][j].number = 2 * board[i-1][j].number;
						board[i-1][j].merged = true;
						board[i][j].number = 0;
						board[i][j].merged = false;
						did_swap = true;
					}
				}
			}
		}
	}
	else if (move == 2)  // Right
	{
		bool did_swap = true;
		while(did_swap)
		{
			did_swap = false;
			for (int i = 0; i < 4; i++)
			{
				for(int j = 3; j > -1; j--)
				{
					if (j == 3)
					{
						
					}
					else if (board[i][j+1].number == 0 && board[i][j].number != 0)
					{
						board[i][j+1] = board[i][j];
						board[i][j].number = 0;
						board[i][j].merged = false;
						did_swap = true;
					}
					else if ((board[i][j+1].number == board[i][j].number) && (board[i][j+1].merged == false && board[i][j].merged == false))
					{
						board[i][j+1].number = 2 * board[i][j+1].number;
						board[i][j+1].merged = true;
						board[i][j].number = 0;
						board[i][j].merged = false;
						did_swap = true;
					}
				}
			}
		}
	}
	else if (move == 3)  // Down
	{
		bool did_swap = true;
		while(did_swap)
		{
			did_swap = false;
			for (int j = 0; j < 4; j++)
			{
				for(int i = 3; i > -1; i--)
				{
					if (i == 3)
					{
						
					}
					else if (board[i+1][j].number == 0 && board[i][j].number != 0)
					{
						board[i+1][j] = board[i][j];
						board[i][j].number = 0;
						board[i][j].merged = false;
						did_swap = true;
					}
					else if ((board[i+1][j].number == board[i][j].number) && (board[i+1][j].merged == false && board[i][j].merged == false))
					{
						board[i+1][j].number = 2 * board[i+1][j].number;
						board[i+1][j].merged = true;
						board[i][j].number = 0;
						board[i][j].merged = false;
						did_swap = true;
					}
				}
			}
		}
	}
}


int main(int argc, char *argv[]) {
	/* Solve the 2048 problem. */
	vector<vector<Tile> > board(4, vector<Tile>(4));
	for (int i = 0; i < 4; i++)
	{
		for (int j = 0; j < 4; j++)
		{
			int this_number;
			cin >> this_number;
			board[i][j].number = this_number;
		}
	}
	int move;
	cin >> move;
	do_move(move, board);
	display_board(board);
}